### PR TITLE
fix(sns): add condition to sns topics

### PR DESCRIPTION
**Pull request by pedrooot**
### Description

It's needed to add new conditions related with organizations. One example is: `"aws:PrincipalOrgID": "<our-org-id>"`
To apply this change we will need to add info for the check and modify the output status using is_condition_block_restrictive_organization

The final approach it's that if the condition has "*" for the "aws:PrincipalOrgID" it won't pass.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
 def is_condition_block_restrictive(
-    condition_statement: dict, source_account: str, is_cross_account_allowed=False
+    condition_statement: dict,
+    source_account: str,
+    is_cross_account_allowed=False,
+    org_id: str = None,
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: Do we need this? Can we not just check if the condition is set?

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py (hunk 1 above)
Comment (pedrooot) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: added org_id since the format for this sort of conditions where like: `"aws:resourceaccount": "<aws_account>"` but for the new condition the org_id it´s needed. This way we ensure that the condition it´s well applied.

File: tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py [hunk 2]
```diff
@@ -226,3 +239,37 @@ def test_topic_public(self):
...
+                sns_topics_not_publicly_accessible,
+            )
+
+            check = sns_topics_not_publicly_accessible()
+            result = check.execute()
+            assert len(result) == 1
+            assert result[0].status == "PASS"
+            assert (
+                result[0].status_extended
+                == f"SNS topic {topic_name} is not public because its policy only allows access from the same account."
```
Comment (MrCloudSec) [inline tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py]: We will have to add another status, if they have an organization in the conditions it has cross-account access for all the accounts inside the org.

Comment (jfagoagas) [reply tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py]: Good catch!

**Comment by MrCloudSec** [comment]
Please @pedrooot, update the PR with the explanation of this fix.

**Comment by pedrooot** [comment]
Done! @sergargar

File: tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py (hunk 2 above)
Comment (mlanders-spectra) [reply tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py]: I'd replace "from the same account" with "from a specific org or specific accounts"

Comment (jfagoagas) [reply tests/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible_test.py]: Agree with @mlanders-spectra.

File: poetry.lock
Comment (jfagoagas) [inline poetry.lock:1]: Could you remove changes in this file?

File: prowler/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible.py [hunk 3]
```diff
@@ -33,13 +33,18 @@ def execute(self):
...
-                            if (
-                                "Condition" in statement
-                                and is_condition_block_restrictive(
-                                    statement["Condition"], sns_client.audited_account
+                            if "Condition" in statement:
+                                account_condition, org_condition = (
+                                    is_condition_block_restrictive(
+                                        statement["Condition"],
+                                        sns_client.audited_account,
+                                        sns_client.audited_org_id,
```
Comment (jfagoagas) [inline prowler/providers/aws/services/sns/sns_topics_not_publicly_accessible/sns_topics_not_publicly_accessible.py]: Where does the `audited_org_id` come from?

I see the point of calling that function with the AWS Account ID since is something we always have but we can not always know the Organization ID audited. In that case something must be passed from config if not set.

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py (hunk 1 above)
Comment (jfagoagas) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: I'm more in favor of just checking if this condition is met. This function just checks if the `condition` block of an IAM policy is restrictive, which means that at least one of the conditions down there is set and with an specific value, not an `*`.

Also, I don't like the need of changing the function result since applies to many other checks.

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -73,7 +73,7 @@ jobs:
       - name: Safety
         if: steps.are-non-ignored-files-changed.outputs.any_changed == 'true'
         run: |
-          poetry run safety check --ignore 70612
+          poetry run safety check --ignore 70612,40459
```
Comment (jfagoagas) [inline .github/workflows/pull-request.yml]: Why is this needed in this PR?

File: docs/tutorials/configuration_file.md [hunk 5]
```diff
@@ -355,6 +356,10 @@ aws:
   # aws.acm_certificates_expiration_check
   days_to_expire_threshold: 7
 
+  # AWS SNS Configuration
+  # aws.sns_topics_not_publicly_accessible
+  organization_id: null
```
Comment (jfagoagas) [inline docs/tutorials/configuration_file.md]: What about several organizations sharing a resource?

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 6]
```diff
@@ -31,6 +36,7 @@ def is_condition_block_restrictive(
             "aws:sourcearn",
             "aws:sourcevpc",
             "aws:sourcevpce",
+            "aws:principalorgid",
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: I think this is no longer needed right?

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 7]
```diff
@@ -41,6 +47,7 @@ def is_condition_block_restrictive(
             "aws:principalaccount",
             "aws:sourcevpc",
             "aws:sourcevpce",
+            "aws:principalorgid",
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: I think this is no longer needed right?

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 8]
```diff
@@ -90,3 +97,76 @@ def is_condition_block_restrictive(
                                 is_condition_valid = True
 
     return is_condition_valid
+
+
+def is_condition_block_restrictive_organization(
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py:100]: Why not to check just if the condition is present at least for now? I mean not to introduce more configuration and logic just to check if a condition is met.

Comment (jfagoagas) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py:100]: What do you think @sergargar?

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 9]
```diff
@@ -1,8 +1,10 @@
-def is_condition_block_restrictive(
-    condition_statement: dict, source_account: str, is_cross_account_allowed=False
+def is_condition_block_restrictive_account(
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: I'd also leave this function with their original name. We can rename that later not to mix things in the PR.

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: .github/workflows/pull-request.yml (hunk 4 above)
Comment (pedrooot) [reply .github/workflows/pull-request.yml]: I'll add it in other PR

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py (hunk 6 above)
Comment (pedrooot) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: True

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py (hunk 7 above)
Comment (pedrooot) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: True

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py (hunk 9 above)
Comment (pedrooot) [reply prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]: Done

File: prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py [hunk 10]
```diff
@@ -90,3 +95,64 @@ def is_condition_block_restrictive(
...
+                }
+
+                if value in condition_statement[condition_operator]:
+                    # values are a list
+                    if isinstance(
+                        condition_statement[condition_operator][value],
+                        list,
+                    ):
+                        is_condition_valid = True
+                        # i
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py]:
```suggestion
```

File: tests/config/config_test.py [hunk 11]
```diff
@@ -262,6 +262,7 @@ def mock_prowler_get_latest_release(_, **kwargs):
     ],
     "check_rds_instance_replicas": False,
     "days_to_expire_threshold": 7,
+    "organization_id": None,
```
Comment (jfagoagas) [inline tests/config/config_test.py]:
```suggestion
```

**Review by jfagoagas — DISMISSED** [review]
That's fantastic 👏

File: prowler/config/config.yaml [hunk 12]
```diff
@@ -283,7 +283,6 @@ aws:
       "controllerManager",
       "scheduler",
     ]
```
Comment (MrCloudSec) [inline prowler/config/config.yaml]:
```suggestion
    ]

```

File: tests/config/fixtures/config.yaml [hunk 13]
```diff
@@ -284,7 +284,6 @@ aws:
         "scheduler",
     ]
```
Comment (MrCloudSec) [inline tests/config/fixtures/config.yaml]:
```suggestion


```

**Review by MrCloudSec — APPROVED** [review]
👑 👑
